### PR TITLE
Fix the barycentre of the secondary subsystem in the rotating-pulsating frame

### PR DESCRIPTION
**Pull request by eggrobin**
#3358.
Fixing the secondary leads to such entertaining equipotentials as the ones shown in #3630 for the Sun-Jool system.
With this change the Sun-Jool equipotentials are as expected. The same applies to Neptune-Sun or Pluto-Sun, though those look a little odd because of perturbations.

Note: On top of #3630.

screenshot4

### CODE REVIEW

File: ksp_plugin/interface_flight_plan.cpp [hunk 1]
```diff
@@ -84,7 +84,7 @@ Burn GetBurn(Plugin const& plugin,
       parameters.primary_index = plugin.CelestialIndexOfBody(
           *barycentric_rotating_reference_frame->primary());
       parameters.secondary_index = plugin.CelestialIndexOfBody(
-          *barycentric_rotating_reference_frame->secondary());
+          *barycentric_rotating_reference_frame->secondaries().front());
```
Comment (pleroy) [inline ksp_plugin/interface_flight_plan.cpp:87]: Maybe check that there is only one?  Or, if there is a convention that `front` is special, comment it.

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -1183,13 +1183,18 @@ Plugin::NewBodySurfaceNavigationFrame(
 }
 
 not_null<std::unique_ptr<PlottingFrame>>
-Plugin::NewRotatingPulsatingPlottingFrame(Index const primary_index,
-                                          Index const secondary_index) const {
+Plugin::NewRotatingPulsatingPlottingFrame(
+    Index const primary_index,
+    std::vector<Index> const secondary_indices) const {
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp:1188]: Should this be a `flat_hash_set`?  If the indices repeat, something is very wrong.

File: ksp_plugin_adapter/reference_frame_selector.cs [hunk 3]
```diff
@@ -704,9 +721,11 @@ internal class ReferenceFrameSelector<ReferenceFrameParameters> : SupervisedWind
         frame_type != old_frame_type ||
         selected_celestial != old_selected_celestial ||
         target_frame_selected != target_frame_was_selected) {
+      // The cast here relies on ReferenceFrameParameters being a class (that
```
Comment (pleroy) [inline ksp_plugin_adapter/reference_frame_selector.cs:724]: Why do we need this change?

File: physics/barycentric_rotating_reference_frame.hpp [hunk 4]
```diff
@@ -55,8 +60,13 @@ class BarycentricRotatingReferenceFrame
...
       not_null<MassiveBody const*> secondary);
 
+  BarycentricRotatingReferenceFrame(
+      not_null<Ephemeris<InertialFrame> const*> ephemeris,
+      not_null<MassiveBody const*> primary,
+      std::vector<not_null<MassiveBody const*>> secondaries);
+
   not_null<MassiveBody const*> primary() const;
-  not_null<MassiveBody const*> secondary() const;
+  std::vector<not_null<MassiveBody const*>> const& secondaries() const;
```
Comment (pleroy) [inline physics/barycentric_rotating_reference_frame.hpp:69]: Same comment regarding a `flat_hash_set`.

File: physics/barycentric_rotating_reference_frame_body.hpp [hunk 5]
```diff
@@ -26,15 +26,24 @@ using namespace principia::quantities::_si;
...
-    not_null<Ephemeris<InertialFrame> const*> ephemeris,
-    not_null<MassiveBody const*> primary,
-    not_null<MassiveBody const*> secondary)
+    BarycentricRotatingReferenceFrame(
+        not_null<Ephemeris<InertialFrame> const*> ephemeris,
+        not_null<MassiveBody const*> primary,
+        not_null<MassiveBody const*> secondary)
+    : BarycentricRotatingReferenceFrame(ephemeris,
+                                        primary,
+                                        std::vector{secondary}) {}
```
Comment (pleroy) [inline physics/barycentric_rotating_reference_frame_body.hpp:35]: Is this constructor used by anything now?

File: physics/rotating_pulsating_reference_frame_body.hpp [hunk 6]
```diff
@@ -2,25 +2,38 @@
...
 template<typename InertialFrame, typename ThisFrame>
 RotatingPulsatingReferenceFrame<InertialFrame, ThisFrame>::
-RotatingPulsatingReferenceFrame(
-    not_null<Ephemeris<InertialFrame> const*> const ephemeris,
-    not_null<MassiveBody const*> const primary,
-    not_null<MassiveBody const*> const secondary)
+    RotatingPulsatingReferenceFrame(
+        not_null<Ephemeris<InertialFrame> const*> const ephemeris,
+        not_null<MassiveBody const*> const primary,
+        not_null<MassiveBody const*> const secondary)
```
Comment (pleroy) [inline physics/rotating_pulsating_reference_frame_body.hpp:22]: Same question, do we use this?

File: physics/barycentric_rotating_reference_frame_body.hpp (hunk 5 above)
Comment (eggrobin) [reply physics/barycentric_rotating_reference_frame_body.hpp:35]: At least the plugin and the test.

File: physics/rotating_pulsating_reference_frame_body.hpp (hunk 6 above)
Comment (eggrobin) [reply physics/rotating_pulsating_reference_frame_body.hpp:22]: At least in a test.

File: physics/rotating_pulsating_reference_frame_body.hpp [hunk 7]
```diff
@@ -127,37 +142,64 @@ not_null<
 RotatingPulsatingReferenceFrame<InertialFrame, ThisFrame>::ReadFromMessage(
     not_null<Ephemeris<InertialFrame> const*> const ephemeris,
     serialization::RotatingPulsatingReferenceFrame const& message) {
+  std::vector<not_null<MassiveBody const*>> secondaries;
+  secondaries.reserve(message.secondary().size());
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply physics/rotating_pulsating_reference_frame_body.hpp:146]: That check should really be in the constructor.

**Comment by eggrobin** [comment]
retest this please

**Comment by eggrobin** [comment]
retest this please

**Comment by eggrobin** [comment]
retest this please

File: ksp_plugin_adapter/interface.cs [hunk 8]
```diff
@@ -140,12 +141,14 @@ internal partial class PlottingFrameParameters : ReferenceFrameParameters {
 
 internal partial struct NavigationFrameParameters : ReferenceFrameParameters {
   public static implicit operator PlottingFrameParameters(NavigationFrameParameters p) {
-    return new PlottingFrameParameters{
+    PlottingFrameParameters result = new PlottingFrameParameters {
```
Comment (pleroy) [inline ksp_plugin_adapter/interface.cs:144]:
```suggestion
    var result = new PlottingFrameParameters{
```

**Pull request merged**